### PR TITLE
Reduce pollution of system directories during test runs

### LINKED ISSUE #48884: Directory /tmp full of QGIS temporary files after ctest run
### What is the bug or the crash?

The /tmp/ directory is filled with temporary files created by QGIS testsuite and not cleaned up on completion.
This is a partial list:
```
dot-qgis
oauth2-cache
QGIS-TEST-gAWuVQ
QGIS-TEST.RnZLfj
srs6.db
```

It would help if at least those files were in a SINGLE directory. Maybe it can be done with a fixture creating such temporary dir..

### Steps to reproduce the issue

Run any test, for example:

    xvfb-run ctest --output-on-failure -R gui_datumtransform

Check the /tmp directory before and after that run

### Versions

Current master as of 6 June 2022, aka https://github.com/qgis/QGIS/commit/ac7b711dea70386490c835daa9b1b2997c682563

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

_No response_

### LINKED ISSUE #50587: Tests create directories under ~/.local/share
### What is the bug or the crash?

Whenever tests are run they create a directory in user dir, so that by the end of `make check` you are left with a lot of directories matching this pattern:
> ls -dl ~/.local/share/qgis* ~/.local/share/test*

### Steps to reproduce the issue

run any test

### Versions

current master

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### LINKED ISSUE #57610: test_core_project test fail if TMPDIR is set to anything but /tmp
### What is the bug or the crash?

The test_core_project test fails if you set TMPDIR to anything but '/tmp' (or unset it, which defaults to '/tmp'.

fails:
```
FAIL!  : TestQgsProject::testAsynchronousLayerLoading() Compared values are not the same
   Actual   (project->mapLayers( true ).count()): 0
   Expected (layersCount - 2)                   : 57
   Loc: [/usr/local/src/qgis/qgis/tests/src/core/testqgsproject.cpp(1098)]
```

### Steps to reproduce the issue

TMPDIR=$PWD/tmp ctest --output-on-failure -R ^test_core_project$

### Versions

current master branch as of commit 92f182568f25b06cb15c5a9d79ea542bda4e9b4d

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

It blocks the ability to reduce filesystem pollution, see GH-57606

### PR DESCRIPTION
**Pull request by strk**
Closes GH-57610
Closes GH-50587
Closes GH-48884

### CODE REVIEW

**Comment by strk** [comment]
For the record: this is the count of files found in the tmp/ directory after `make check`:
{{{
find tmp/ -type f | wc -l
3916
}}}

All those will now at least be under a single directory, to make their removal easier (with this PR)

**Comment by strk** [comment]
The failure in this PR is due to a bug in a test which is fixed by GH-57612 -- once that is merged this can be rebased

**Comment by lnicola** [comment]
Hey @strk, this looks like a pretty trivial PR, but nobody seems to be available to merge it.

But it doesn't look obviously wrong to me. Do you want me to file an identical PR that you could approve, working around the slow review times? It would mean bypassing the merge rules, but if nobody complains maybe that's fine.

**Review by elpaso — APPROVED** [review]

**Comment by nyalldawson** [comment]
@strk I need to revert this -- it's broken my build and causes running tests from a non-installed build to segfault.

**Comment by strk** [comment]
@nyalldawson could you file a separate ticket and add a stacktrace ? A segfault upon specifying a custom tmpdir is most likely a missing check in a testcase

File: tests/src/core/testqgslabelingengine.cpp [hunk 1]
```diff
@@ -4182,7 +4182,7 @@ void TestQgsLabelingEngine::labelingResultsWithCallouts()
   job.start();
   job.waitForFinished();
 
-  job.renderedImage().save( QStringLiteral( "/tmp/renderer.png" ) );
+  job.renderedImage().save( QString( "%s/renderer.png" ).arg( QDir::tempPath() ) );
```
Comment (nyalldawson) [inline tests/src/core/testqgslabelingengine.cpp:4185]: This isn't valid QString syntax. See https://doc.qt.io/qt-6/qstring.html#arg.

**Comment by nyalldawson** [comment]
@strk 


```
nyall@nyall-framework:~/dev/build-qgis$ ctest -R GdalProvider\$ -V
UpdateCTestConfiguration  from :/home/nyall/dev/build-qgis/DartConfiguration.tcl
Parse Config file:/home/nyall/dev/build-qgis/DartConfiguration.tcl
 Add coverage exclude regular expressions.
 Add coverage exclude: /CMakeFiles/CMakeTmp/
UpdateCTestConfiguration  from :/home/nyall/dev/build-qgis/DartConfiguration.tcl
Parse Config file:/home/nyall/dev/build-qgis/DartConfiguration.tcl
Test project /home/nyall/dev/build-qgis
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added setup test logGitStatus required by fixture SOURCETREE
Added cleanup test checkGitStatus required by fixture SOURCETREE
Added 2 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start   1: logGitStatus

1: Test command: /home/nyall/dev/QGIS/scripts/chkgitstatus.sh "log"
1: Working Directory: /home/nyall/dev/build-qgis
1: Test timeout computed to be: 1500
1/3 Test   #1: logGitStatus .....................   Passed    0.05 sec
test 496
    Start 496: PyQgsGdalProvider

496: Test command: /usr/bin/cmake "-P" "/home/nyall/dev/build-qgis/tests/src/python/PyQgsGdalProvider.cmake"
496: Working Directory: /home/nyall/dev/build-qgis/tests/src/python
496: Environment variables: 
496:  TMPDIR=/home/nyall/dev/build-qgis/tmp
496:  QGIS_CUSTOM_CONFIG_PATH=/home/nyall/dev/build-qgis/tmp/config
496: Test timeout computed to be: 1500
496: export =
496: export PYTHONPATH=/home/nyall/dev/build-qgis/output/python:/home/nyall/dev/build-qgis/output/python/plugins:/home/nyall/dev/QGIS/tests/src/python:
496: -- Running  /usr/bin/python3.12 /home/nyall/dev/QGIS/tests/src/python/test_provider_gdal.py
496: src/core/providers/qgsproviderregistry.cpp:372 : (init) [339ms] Loaded 33 providers (OAPIF;WFS;arcgisfeatureserver;arcgismapserver;arcgisvectortileservice;cesiumtiles;copc;delimitedtext;ept;gdal;gpx;grass;grassraster;mbtilesvectortiles;mdal;memory;mesh_memory;mssql;ogr;pdal;postgres;postgresraster;sensorthings;spatialite;tiledscene;vectortile;virtual;virtualraster;vpc;vtpkvectortiles;wcs;wms;xyzvectortiles) 
496: QFile::remove: Empty or null file name
496: /usr/lib64/python3.12/site-packages/osgeo/gdal.py:312: FutureWarning: Neither gdal.UseExceptions() nor gdal.DontUseExceptions() has been explicitly called. In GDAL 4.0, exceptions will be enabled by default.
496:   warnings.warn(
496: GDAL ERROR 1: Unable to compute a transformation between pixel/line and georeferenced coordinates for testInt8.tif. There is no affine transformation and no GCPs. Specify transformation option SRC_METHOD=NO_GEOTRANSFORM to bypass this check.
496: GDAL ERROR 1: Unable to compute a transformation between pixel/line and georeferenced coordinates for testInt8.tif. There is no affine transformation and no GCPs. Specify transformation option SRC_METHOD=NO_GEOTRANSFORM to bypass this check.
496: .....src/core/qgsmessagelog.cpp:39 : (logMessage) [121ms] 2024-06-27T12:57:07 WCS[1] Download of capabilities failed: Protocol "" is unknown
496: src/core/qgsmessagelog.cpp:39 : (logMessage) [0ms] 2024-06-27T12:57:07 WCS[1] Download of capabilities failed: Protocol "" is unknown
496: src/providers/wcs/qgswcscapabilities.cpp:326 : (describeCoverage) [0ms] coverage not found
496: .Could not create temporary copy
496: CMake Error at PyQgsGdalProvider.cmake:23 (message):
496:   Test failed: Subprocess aborted
```

File: tests/src/core/testqgslabelingengine.cpp (hunk 1 above)
Comment (strk) [reply tests/src/core/testqgslabelingengine.cpp:4185]: Good catch! Handled in GH-57934

**Pull request merged**